src/rust.rs

```diff
@@ -330,10 +330,14 @@ impl<T: Copy> MutableHandle<T> {
 
 impl<T> Drop for Rooted<T> {
     fn drop(&mut self) {
+        if self.stack.is_null() {
```

Comment:
Heh, I should have tried to write this. In my defence I was on a bus while testing.

I think this won't work, since rust fills the dropped objects (even those marked with `#[unsafe_no_drop_flag]`) with `0x1d` (see https://github.com/rust-lang/rust/blob/040a77f772d7693598499a161f53ed230fb61c52/src/libcore/mem.rs#L539).

A quick test with this patch applied over #227 yields:

```
(gdb) run
Starting program: /home/emilio/projects/moz/rust-mozjs/target/debug/js-d2b536494d16a8fc 
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".

running 2 tests
[New Thread 0x7ffff6b31700 (LWP 17770)]
[New Thread 0x7ffff5fff700 (LWP 17771)]
[New Thread 0x7ffff57ff700 (LWP 17772)]
[New Thread 0x7ffff4fff700 (LWP 17773)]
[New Thread 0x7ffff7fc0700 (LWP 17774)]
[New Thread 0x7ffff7f3f700 (LWP 17775)]
[New Thread 0x7ffff7ebe700 (LWP 17776)]
[New Thread 0x7ffff6930700 (LWP 17777)]
[New Thread 0x7ffff68af700 (LWP 17778)]
[New Thread 0x7ffff5dfe700 (LWP 17779)]
[New Thread 0x7ffff5d7d700 (LWP 17780)]
[New Thread 0x7ffff5cfc700 (LWP 17781)]
[New Thread 0x7ffff55fe700 (LWP 17782)]
[New Thread 0x7ffff557d700 (LWP 17783)]
[New Thread 0x7ffff54fc700 (LWP 17784)]
[New Thread 0x7ffff4dfe700 (LWP 17785)]

Program received signal SIGSEGV, Segmentation fault.
[Switching to Thread 0x7ffff57ff700 (LWP 17772)]
0x000055555563bc7c in js::rust::Rooted<T>.Drop::drop (self=0x7ffff57fe980) at src/rust.rs:337
337	            assert!(*self.stack == mem::transmute(&*self));
(gdb) p self.stack
$1 = (struct Rooted<*mut libc::c_void> **) 0x1d1d1d1d1d1d1d1d
(gdb) p self
$2 = (struct Rooted<*mut js::jsapi_linux_64::JSObject> *) 0x7ffff57fe980
(gdb) p self._base
$3 = {_phantom0 = {<No data fields>}}
(gdb) p self.prev
$4 = (struct Rooted<*mut libc::c_void> *) 0x1d1d1d1d1d1d1d1d
(gdb) p self.ptr
$5 = (union JSObject *) 0x1d1d1d1d1d1d1d1d
```

We might want to check for `0x1d1d1d` (I just added a commit to #226 that does it, and works) , but it feels a bit too hacky...